File: src/main/java/com/realyoungk/sdi/entity/UserEntity.java
package com.realyoungk.sdi.entity;

import com.realyoungk.sdi.model.CalendarType; // import 추가
import com.realyoungk.sdi.model.LunarMonthType;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.util.UUID;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
@Table(name = "users")
public class UserEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, updatable = false)
    private String publicId;

    @Column(nullable = false, length = 50)
    private String name;

    @Column(length = 20)
    private String phoneNumber;

    private LocalDate birthday;

    @Enumerated(EnumType.STRING)
    private CalendarType calendarType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private LunarMonthType lunarMonthType = LunarMonthType.NORMAL;


    @Column(length = 100)
    private String teamName;

    @PrePersist
    protected void onCreate() {
        if (this.publicId == null) {
            this.publicId = UUID.randomUUID().toString();
        }
    }

    @Builder
    public UserEntity(String name, String phoneNumber, LocalDate birthday, CalendarType calendarType, LunarMonthType lunarMonthType, String teamName) {
        this.name = name;
        this.phoneNumber = phoneNumber;
        this.birthday = birthday;
        this.calendarType = calendarType;
        this.lunarMonthType = (calendarType == CalendarType.SOLAR) ? LunarMonthType.NORMAL : lunarMonthType;
        this.teamName = teamName;
    }
}

File: src/main/kotlin/com/realyoungk/sdi/entity/UserEntity.kt
package com.realyoungk.sdi.entity

import com.realyoungk.sdi.enums.CalendarType
import com.realyoungk.sdi.enums.LunarMonthType
import jakarta.persistence.*
import java.time.LocalDate
import java.util.*

@Entity
@Table(name = "users")
open class UserEntity protected constructor() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    @Column(nullable = false, unique = true, updatable = false)
    var publicId: String? = null
        protected set

    @Column(nullable = false, length = 50)
    var name: String? = null
        protected set

    @Column(length = 20)
    var phoneNumber: String? = null
        protected set

    var birthday: LocalDate? = null
        protected set

    @Enumerated(EnumType.STRING)
    var calendarType: CalendarType? = null
        protected set

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var lunarMonthType: LunarMonthType = LunarMonthType.NORMAL
        protected set

    @Column(length = 100)
    var teamName: String? = null
        protected set

    @PrePersist
    protected fun onCreate() {
        if (publicId == null) {
            publicId = UUID.randomUUID().toString()
        }
    }

    companion object {
        @JvmStatic
        fun builder(): UserEntityBuilder = UserEntityBuilder()
    }

    class UserEntityBuilder {
        private var name: String? = null
        private var phoneNumber: String? = null
        private var birthday: LocalDate? = null
        private var calendarType: CalendarType? = null
        private var lunarMonthType: LunarMonthType? = null
        private var teamName: String? = null

        fun name(name: String) = apply { this.name = name }
        fun phoneNumber(phoneNumber: String?) = apply { this.phoneNumber = phoneNumber }
        fun birthday(birthday: LocalDate?) = apply { this.birthday = birthday }
        fun calendarType(calendarType: CalendarType?) = apply { this.calendarType = calendarType }
        fun lunarMonthType(lunarMonthType: LunarMonthType?) = apply {
            this.lunarMonthType = lunarMonthType
        }

        fun teamName(teamName: String?) = apply { this.teamName = teamName }

        fun build(): UserEntity {
            val entity = UserEntity()
            entity.name = name
            entity.phoneNumber = phoneNumber
            entity.birthday = birthday
            entity.calendarType = calendarType
            entity.lunarMonthType = if (calendarType == CalendarType.SOLAR) {
                LunarMonthType.NORMAL
            } else {
                lunarMonthType ?: LunarMonthType.NORMAL
            }
            entity.teamName = teamName
            return entity
        }
    }
}